src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/KeyboardEventSender.cs

```diff
@@ -12,95 +13,102 @@ namespace RegressionGames.StateRecorder
-                    deltaStateEvent = newEvent;
                 }
+                
+                InputSystem.QueueEvent(eventPtr);
+            }
+        }
 
-                logList.Add(valueTuple);
-                keys.Add(key);
-                var inputControl = keyboard.allControls
-                    .FirstOrDefault(a => a is KeyControl kc && kc.keyCode == key) ?? keyboard.anyKey;
-                if (inputControl != null)
+        private static void QueueTextEvent(int replaySegment, Key key)
+        {
+            // send a text event so that 'onChange' text events fire
+            // convert key to text
+            if (KeyboardInputActionObserver.KeyboardKeyToValueMap.TryGetValue(key, out var possibleValues))
+            {
+                var value = _isShiftDown ? possibleValues.Item2 : possibleValues.Item1;
+                if (value == 0x00)
                 {
-                    inputControl.WriteValueIntoEvent(upOrDown == KeyState.Down ? 1f : 0f, deltaStateEvent.Value);
-
-                    if (upOrDown == KeyState.Down)
-                    {
-                        // send a text event so that 'onChange' text events fire
-                        // convert key to text
-                        if (KeyboardInputActionObserver.KeyboardKeyToValueMap.TryGetValue(((KeyControl)inputControl).keyCode, out var possibleValues))
-                        {
-                            var value = _isShiftDown ? possibleValues.Item2 : possibleValues.Item1;
-                            if (value == 0x00)
-                            {
-                                RGDebug.LogError($"Found null value for keyboard input {key}");
-                                return;
-                            }
-
-                            var inputEvent = TextEvent.Create(Keyboard.current.deviceId, value, time);
-                            RGDebug.LogInfo($"({replaySegment}) Sending Text Event for char: '{value}'");
-                            InputSystem.QueueEvent(ref inputEvent);
-                        }
-                    }
+                    RGDebug.LogError($"Found null value for keyboard input {key}");
+                    return;
                 }
 
+                var inputEvent = TextEvent.Create(Keyboard.current.deviceId, value, InputState.currentTime);
+                RGDebug.LogInfo($"({replaySegment}) Sending Text Event for char: '{value}'");
+                InputSystem.QueueEvent(ref inputEvent);
+            }
+        }
+
+        /**
+         * Allows updating multiple keys in the same event. 
+         */
+        public static void SendKeysInOneEvent(int replaySegment, List<(Key, KeyState)> keyStates)
+        {
+            #if ENABLE_LEGACY_INPUT_MANAGER
+            SendKeysInOneEventLegacy(keyStates);
+            #endif
+            
+            RGDebug.LogInfo($"({replaySegment}) Sending Multiple Key Event: [" + string.Join(", ", keyStates.Select(a => a.Item1 + ":" + a.Item2).ToArray()) + "]");
```

Comment:
(non-blocking) Does that mean legacy input systems sent inputs here and then again down below? Curious why this doesn't return here.

Reply:
The code that follows afterwards is the code for sending the key events down the new input system - we still want to send events down both systems in case the game is using both the new and old input systems.

The actual implementation of `SendKeysInOneEventLegacy` just sends multiple individual events through `RGLegacyInputWrapper`. There isn't really any need to group the events for the legacy input support in the same way as we do for the input system